pkg/platform/kvmlibvirt/template.go

```diff
@@ -0,0 +1,190 @@
+
+  {{- if .Config.PodCidr }}
+  pod_cidr = {{.Config.PodCidr }}
+  {{- end }}
+  {{- if .Config.ServiceCidr }}
+  service_cidr = {{.Config.ServiceCidr }}
+  {{- end }}
+
+  {{- if .Config.ClusterDomainSuffix }}
+  cluster_domain_suffix = {{.Config.ClusterDomainSuffix }}
+  {{- end }}
+
+  enable_reporting   = {{.Config.EnableReporting}}
+  enable_aggregation = {{.Config.EnableAggregation}}
+
+  {{- if .Config.ControllerCLCSnippets }}
+  controller_clc_snippets = {{ StringsJoin .Config.ControllerCLCSnippets "\", \"" | printf "[%q]" }}
+  {{- end }}
+
+  disable_self_hosted_kubelet = {{ .Config.DisableSelfHostedKubelet }}
+
+  {{- if .Config.KubeAPIServerExtraFlags }}
+  kube_apiserver_extra_flags = {{ StringsJoin .Config.KubeAPIServerExtraFlags "\", \"" | printf "[%q]" }}
+  {{- end }}
+
+  {{- if .Config.CertsValidityPeriodHours }}
+  certs_validity_period_hours = {{.Config.CertsValidityPeriodHours}}
+  {{- end }}
+
+ # Enable TLS Bootstrap
+   enable_tls_bootstrap = {{ .Config.EnableTLSBootstrap }}
+  worker_bootstrap_tokens = [
+    {{- range $index, $pool := .Config.WorkerPools }}
+    module.worker-pool-{{$index}}.worker_bootstrap_token,
+    {{- end }}
+  ]
+}
+
+{{ range $index, $pool := .Config.WorkerPools }}
+module "worker-pool-{{ $index }}" {
+  source = "../terraform-modules/kvm-libvirt/flatcar-linux/kubernetes/workers"
+
+  ssh_keys = {{ StringsJoin $.Config.SSHPubKeys "\", \"" | printf "[%q]" }}
+  machine_domain        = "{{$.Config.MachineDomain}}"
+  cluster_name          = "{{$.Config.ClusterName}}"
+  {{- if $.Config.ClusterDomainSuffix }}
+  cluster_domain_suffix = {{$.Config.ClusterDomainSuffix }}
+  {{- end }}
+  {{- if $.Config.ServiceCidr }}
+  service_cidr = {{$.Config.ServiceCidr }}
+  {{- end }}
+
+  libvirtpool           = module.kvm-libvirt-{{$.Config.ClusterName}}.libvirtpool
+  libvirtbaseid         = module.kvm-libvirt-{{$.Config.ClusterName}}.libvirtbaseid
+  kubeconfig            = module.kvm-libvirt-{{$.Config.ClusterName}}.kubeconfig
+
+  ca_cert              = module.kvm-libvirt-{{ $.Config.ClusterName }}.ca_cert
+  apiserver            = module.kvm-libvirt-{{ $.Config.ClusterName }}.apiserver
+  enable_tls_bootstrap = {{ $.Config.EnableTLSBootstrap }}
+
+  pool_name             = "{{ $pool.Name }}"
+  worker_count          = "{{ $pool.Count}}"
+
+  {{- if $pool.VirtualCPUs }}
+  virtual_cpus = {{ $pool.VirtualCPUs }}
+  {{- end }}
+
+  {{- if $pool.VirtualMemory }}
+  virtual_memory {{ $pool.VirtualMemory }}
+  {{- end }}
+
+  {{- if $pool.Labels }}
+  labels = {{ $pool.Labels }}
+  {{- end }}
+
+  {{- if $pool.CLCSnippets }}
+  clc_snippets = {{ StringsJoin $pool.CLCSnippets "\", \"" | printf "[%q]" }}
+  {{- end }}
+}
+{{- end }}
+
+terraform {
+  required_providers {
+    libvirt = {
+			source = "dmacvicar/libvirt"
+			uri     = "qemu:///system"
+			version = "0.6.2"
+    }
+    ct = {
+			source  = "poseidon/ct"
+			version = "0.6.1"
+    }
+  }
+}
+
+provider "libvirt" {
+	uri     = "qemu:///system"
```

Comment:
Just started setting up Terraform with existing libvirtd cluster (not lokomotive) and noticed this PR.

For those of us with a home 'server' it would be great if we can pass the kvm host through cluster configuration, perhaps along the lines of:
```
  {{- if .Config.QemuURI }}
  uri     = {{.Config.QemuURI }}
  {{else}} 
  uri     = "qemu:///system"
  {{- end}}
```

Reply:
To use a remote location? Maybe it's not easy to get this works with the image volume pool, or?

Reply:
Providing a:
1. external file eg: `https://example.com/flat-car.img` results in terraform runner (laptop) downloading it and then copying across the network to the server. I think this is reasonable.
2. laptop local file eg: `/path/to/flat-car.img` results in terraform runner copying across network to the server. Also reasonable.

It is unintuitive but I don't think there's an alternative - see: https://github.com/dmacvicar/terraform-provider-libvirt/issues/299#issuecomment-419676846

I currently supply a path to images that's actually a NFS mount from the 'server' which is kind of gross (looping data) but works in the sense that I don't need to keep the image on my laptop. :sweat_smile: 

```
# Qcow2 base OS volume that can be shared by VM's (domain's)
resource "libvirt_volume" "os_base" {
  name   = "${var.guest_hostname}-os_base"
  source = "/mnt/nfs/path/to/ubuntu-focal-server-cloudimg-amd64.img"
}

# OS volume per guest VM
resource "libvirt_volume" "os" {
  count          = var.guest_count
  name           = "${var.guest_hostname}-${format("%02d", count.index)}-os.qcow2"
  base_volume_id = libvirt_volume.os_base.id
  pool           = var.guest_pool_name
  size           = var.os_volume_size
}
```

Reply:
Please let me know if I misunderstood the question.
If the concept is amenable, perhaps we could add a line to markdown file saying that image sources will be fetched by lokoctl executor and copied to remote kvm host?

Maybe I build a fork of this PR with the change and do some testing first. :)
Then I can do a small PR once this one merged or just post here that it works as suspected if this is still being iterated on?

Reply:
I started building this PR and trying to run it against my remote kvm server.

On top of a couple of string quotes suggested here the changes are: https://github.com/kskewes/lokomotive/commit/6246cc76e8e7287759023565e5fc000677aa4266
It creates all of the libvirt domains and things but the default NAT networking doesn't enable access from lokoctl runner to hosts. It really needs to be a bridge network for me. I'll have another go later and can iterate in a fork.